### DISCUSSION TITLE
Flexible background patterns

### DISCUSSION THREAD

**Discussion by letsfindaway** [letsfindaway/OpenBoard#165] [Ideas]
Several users expressed their interest in more background patterns. recently Seyes was added to OpenBoard. But as the number of patterns in the background palette is limited, another pattern has to vanish.

I have now started work for more flexible background patters described in an XML file and want to use this discussion to collect feedback and further ideas.

- See the PR at https://github.com/OpenBoard-org/OpenBoard/pull/888
- and the Wiki page at [Flexible background patterns](https://github.com/letsfindaway/OpenBoard/wiki/Flexible-background-patterns).

**Comment by sebojolais** [comment]
1. First impression. 
It is an awesome feature, very smart, with only good points and no reserve. At first, I was affraid about compatibility issues but you manage all of them.

3. First feedbacks about build issues on linux ubuntu 22.04:
- With `Qt 5.15.6`, I need to add the include of optional in `UBBackgroundRuling.h` in order to finish the build : 
`#include <optional>`
- With `Qt 6.4`, I got some small issues about string conversion. Did you test it with `Qt 6` ?
- Just a comment for other readers, the PR works with `CMake` only currently.

4. Why do you search the background XMLs in a folder called `template` ?
In other words : Why using this folder: 
`/home/.local/share/OpenBoard/template/background/`
instead of a simplier:
`/home/.local/share/OpenBoard/background/`

**Reply by letsfindaway** [reply]
> 1. First impression.
>        It is an awesome feature, very smart, with only good points and no reserve. At first, I was affraid about compatibility issues but you manage all of them.

Thank you! Glad to hear this.
 
> 2. First feedbacks about build issues on linux ubuntu 22.04:
> 
> * With `Qt 5.15.6`, I need to add the include of optional in `UBBackgroundRuling.h` in order to finish the build :
>       `#include <optional>`

Thanks for pointing this out. For me it worked without this, but as I am using this feature I should also explicitly include it.

> * With `Qt 6.4`, I got some small issues about string conversion. Did you test it with `Qt 6` ?

Did not test with Qt 6, but will do.

> * Just a comment for other readers, the PR works with `CMake` only currently.

I tried to fix this with my latest forced push, but did not test it. Will also do this and make it work with `qmake` and `cmake`.

> 4. Why do you search the background XMLs in a folder called `template` ?
>        In other words : Why using this folder:
>        `/home/.local/share/OpenBoard/template/background/`
>        instead of a simplier:
>        `/home/.local/share/OpenBoard/background/`

You are right. There is no need to create this additional hierarchy.

Thanks for your feedback! BTW: I'm missing number 3. in your list ;)

Note that I have a few days off now. Will start working on this next week.

**Reply by sebojolais** [reply]
> I'm missing number 3. in your list ;)

I did a typo ! The number 4 is 3. ;-)

Does a PR with the support of Qt6 help you ? 
Here it is ! #166

**Reply by letsfindaway** [reply]
Just  merged your PR and pushed a commit removing the extra `template` directory. I think this conversation is now resolved.

**Reply by letsfindaway** [reply]
I also tested and it should now also work with `qmake`.

**Comment by sebojolais** [comment]
GUI
------
On the background selector dialog, instead of `Switch background color` I prefer `Dark background` or `Dark board mode`. It is shorter and clearer for me.

**Reply by letsfindaway** [reply]
But then I would have to change the label to "Light background" when I'm in dark mode? I think "switch" is more neutral.

**Reply by sebojolais** [reply]
Ok, I see what you mean. I do not interpret the Gui in the same way.
There is an `OFF` state:
```
 ______
((_)___)
```

And an `ON` state:
```
 ______
(___(_))
```
So in my mind, the _Dark board mode_ is `OFF` or `ON` with no label changement.
But no idea how the majority of users will interpret it ...

**Reply by letsfindaway** [reply]
Now I see what you mean ;) Any smartphone user immediately recognizes this kind of switch and associates the left position with "off" and the right position with "on". Ok, let's keep this point open and see what others think how an intuitive UI should look like and what the best label would be.

**Comment by sebojolais** [comment]
Background drawing issue ?
------
It seems there is an issue if we set a border for 90° lines. For the Seyes background, the vertical lines should start with the margin to the right direction and no vertical line at the left of the margin. (https://en.wikipedia.org/wiki/File:Page_blanche_A5_reglure_Seyes.svg) So I try to modify the XML accordingly. Have e look below:
```

<bg:background xmlns:bg="http://openboard.org/template/background/">
    <bg:uuid>38b45f18-f6a5-4336-ae91-bedba3a2442c</bg:uuid>
    <bg:description xml:lang="en">Seyes</bg:description>
    <bg:attributes ruled="true"/>
    <bg:linegroup>
        <bg:angle>0</bg:angle>
        <bg:spacing>20</bg:spacing>
        <bg:line>
            <bg:offset>0</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#8e7cc3</bg:onDark>
                <bg:onLight>#8e7cc3</bg:onLight>
            </bg:color>
        </bg:line>
        <bg:line>
            <bg:offset>5</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#996fa8dc</bg:onDark>
                <bg:onLight>#996fa8dc</bg:onLight>
            </bg:color>
        </bg:line>
        <bg:line>
            <bg:offset>10</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#996fa8dc</bg:onDark>
                <bg:onLight>#996fa8dc</bg:onLight>
            </bg:color>
        </bg:line>
        <bg:line>
            <bg:offset>15</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#996fa8dc</bg:onDark>
                <bg:onLight>#996fa8dc</bg:onLight>
            </bg:color>
        </bg:line>
    </bg:linegroup>
    <bg:linegroup>
        <bg:angle>90</bg:angle>
        <bg:spacing>20</bg:spacing>
        <bg:origin>topleft</bg:origin>
        <bg:line>
            <bg:offset>0</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#8e7cc3</bg:onDark>
                <bg:onLight>#8e7cc3</bg:onLight>
            </bg:color>
        </bg:line>
        <bg:border>
            <bg:left>0</bg:left>
        </bg:border>
    </bg:linegroup>
    <bg:linegroup>
        <bg:angle>90</bg:angle>
        <bg:origin>topleft</bg:origin>
        <bg:line>
            <bg:offset>20</bg:offset>
            <bg:width>2</bg:width>
            <bg:color>
                <bg:onDark>#ff0000</bg:onDark>
                <bg:onLight>#ff0000</bg:onLight>
            </bg:color>
        </bg:line>
    </bg:linegroup>
</
```bg:background>

I get some strange not really vertical lines: 
![Screenshot_20240215_124314](https://github.com/letsfindaway/OpenBoard/assets/5416255/9e7afeff-87dd-48cc-b637-a1d5d3062b6b)

**Reply by letsfindaway** [reply]
Indeed, this is strange! I will create an issue from this comment. Actually a border is probably not correctly handled if a pattern contains any non-horizontal lines.

**Reply by letsfindaway** [reply]
See #167. Please follow-up this topic in the referenced issue.
